### ISSUE TITLE
Drop of duplicate block does not behave consistently

### ISSUE THREAD

**Issue by kbakalar** [jmoenig/Snap--Build-Your-Own-Blocks#1270]
When you left click on a block and select duplicate from the pull down menu, a duplicate of the selected block (and its children) is attached to the hand. If the duplicate is dropped in a window other than a script window, the behavior differs depending on the the location of the original.  If the original was on the script pane outside of any block editor window, the duplicate is stepped into the trash at the left.  However, if the original was on the script inside a block editor window, the duplicate is deposited where it is dropped (for example, on top of the canvas window). 

The behavior in the second case should match the behavior in the first case.

**Comment by jmoenig** [comment]
Not sure if I understand this right. What *should* happen (at least what I wanted it to do) is that if something is rejected by a target pane it slides back to its former situation, unless it doesn't have one, in which case it gets deleted. So, if you drag a block from the palette and drop it onto the stage it slides back to the palette where it gets deleted. Similarly if you use the context menu to duplicate a block in the scripting area and drop it onto the stage it, too, slides "back" to the palette where it gets deleted (only the duplicate). Can you confirm this behavior? Thanks!

**Comment by kbakalar** [comment]
Yes, I confirm the behavior you describe for your example.  But try this instead. Use the context menu to duplicate a block residing in the _block editor_ window rather than the script pane.  The behavior is not the same as your case, and according to your rule it should be the same.  The duplicate does not slide back to the palette, and it is not deleted. The duplicate becomes a direct child of the WorldMorph.

To reproduce:
1. Open snap.berkeley.edu/init
2. Create a new block with arbitrary name and type
3. Clone an arbitrary block A from the palette and drag it inside the Block Editor.
5. Left click on A, choose "duplicate" from the drop down menu.
6. Drag and drop the copy of A to the "stage" window.
Result:  The duplicate appears over the "stage" window.  The duplicate is now a direct child of WorldMorph.

**Comment by jmoenig** [comment]
Yes, now I see what you mean, excellent catch, thanks! Did you discover any other inconsistencies?

**Comment by kbakalar** [comment]
I have been working on #1130  by way of becoming familiar with the project.  The problem from the point of view of the use model is to decide what constitutes a "dirty" block editor window that needs an Apply. The hard part from the point of view of the internals is to add logic that supports the ideal use model with the fewest extraneous false positives. A false negative is, of course, never tolerable. After trying several strategies, I am now leaning towards the conclusion that any mouseup or mousedown (or their keyboard or touchscreen equivalents) in the confines of the Block Editor ScriptMorph "dirties" the block. This captures some false positives, but it has the virtue of being dead simple to implement. The false positives are not terribly annoying because they are edge cases, not visually confusing, and in any case can be dismissed with a click of the Apply button.

Which brings me to your question.  With regard to script editing, blocks residing inside a block editor window are treated differently from a reactive script residing directly in the scripts pane (e.g., one with a "when clicked" hat).  Changes to a script in the script pane are immediately active. Change to a script in the block editor window are not active until after an Apply. What use case justifies the difference?

**Comment by brianharvey** [comment]
> What use case justifies the difference?

I think this is my fault, so I'll answer this part. :-)  I often (in text languages too) start some modification to a procedure and realize halfway in that what I'm trying to do isn't going to work.  I then want to revert to my last saved version.  That's why I like having to say explicitly that I want my changes saved.

Maybe I'd be happy if the block editor remembered the state of the block as it was when the editor opened, and the editor's context menu had a "revert" option.

**Comment by kbakalar** [comment]
Your answer begs the question. Why is the case any different for the reactive script lying directly in the script pane? I might well want to revert to the last working version in this case too.

**Comment by brianharvey** [comment]
I suppose that's true in principle.  But my instinct is to put anything at all complicated in a procedure.  Toplevel scripts are, like, 
when flag clicked
  pen down
  my custom block

**Comment by brianharvey** [comment]
Hey, you're another lefty!  :-)

I vote no on your rather ponderous, imho, suggestion for reverting toplevel scripts.  If I'm going to do something in advance to enable reversion, I'll just duplicate the whole script and comment it out, not use the block editor.

> That doesn't mean we should impose it on others, especially beginners, who have enough to deal with already.

I object to "impose" here.  Beginners are welcome to just close the block editor when done editing.  And I don't think most beginners expect using the block editor to be *exactly* like using a toplevel script.  The natural thing for a beginner, I think, isn't to edit a procedure while it's running; I certainly don't do that!  I use "Apply" for debugging, allowing me to see (not edit) the code while it's running.  That's a feature for intermediate users, not raw beginners.

**Comment by cycomachead** [comment]
I think we're just coming to the conclusion that we need normal "Undo" support, no? #78 #872 

> The natural thing for a beginner, I think, isn't to edit a procedure while it's running; I certainly don't do that!

I'm not so sure. I've seen lots of students do this over time, but I can't say how far along they would be in their learning. Snap<em>!</em>'s (and Scratch's) reactiveness make this something cool to do. Once users realize this works in normal scripts, I think it _is_ natural to expect they work on blocks. (Especially if you only think of custom blocks as a way to "hide" or "condense" code rather than as all the things that functions / procedures entail.)

**Comment by kbakalar** [comment]
Perhaps extending the revert-style block editor to top-level scripts is a little too much, but using it for command, reporter and predicate blocks seems appealing. #1130 tries to solve the same problem (forgetting to save), but it isn't as clean. 

I think the revert-style block editor is orthogonal to Undo (#78) and can live comfortably with it. Quoting from the note above and extending for clarity and completness, here is the spec:

> For purposes of this definition, the _current script_ associated with a block is the script that will be interpreted at run time when the block is invoked. The revert-style Block Editor is defined as follows. An "**edit**" context menu selection from a block or an "**Okay**" selection in the Make a Block dialog opens the Block Editor, displays the current script and caches a backup. The Block Editor has  "**Revert**" and "**Close**" buttons. Modifications in the window modify the current script. "**Revert**" causes the backup to replace the current script. "**Close**" closes the window and deletes the backup. 

Here is the caveat. The proposal modifies a heavily used feature in way that is not backward compatible. Do you think it's worth opening a separate issue on the topic to gather more intelligence?

**RE:** my ponderous proposal. 

I wasn't thinking about modifying the code while the interpreter is running.  I meant only to say that when a reactive script is promoted to a revert-style block editor window it is interpreted just as if it were still directly on the script pane. 

There is external evidence that students tend to write non-trivial reactive scripts on the script pane:  #78 #872.

**Comment by kbakalar** [comment]
Getting back to the initial problem: it goes away if the condition in block .js  line 2362 of  BLockMorph.userMenu addItem "duplicate" is relaxed to synthesize a "grab" for blocks duplicated from a Block Editor as well as the scripting pane:

        2362    ide = myself.parentThatIsA(IDE_Morph) || myself.parentThatIsA(WorldMorph).children[0]; 

This also seems to work, but it makes the assumption that the ide is always available:

function () {
            myself.fullCopy().pickUp(world);
            var palette = world.children[0].palette;
            world.hand.grabOrigin = {
                origin: palette,
                position: palette.center()
            }
        }
